### ISSUE TITLE
False negative when deep expression operator combined with ellipsis operator

### ISSUE THREAD

**Issue by mschwager** [semgrep/semgrep#1804]
I'm looking for a function call that could be in an `if` statement, or it could not be, so I'm using the deep expression operator. I'm then looking for a subsequent function call using the same argument. My searches for either function call on its own each work, but combined they don't work.

- Deep expression operator works: https://dev.semgrep.dev/s/X2Rl/
- Subsequent function call works: https://dev.semgrep.dev/s/jxWZ/
- Combined they don't work: https://dev.semgrep.dev/s/1LDb/

Am I missing anything, or is this a known shortcoming of the deep expression operator? cc @aryx.

Note that a double deep expression operator also fails: https://dev.semgrep.dev/s/9Rpk/.

**Comment by mjambon** [comment]
Simplified example, target python code:
```python
if a:
    pass
b
```

Pattern that fails but was expected to match https://dev.semgrep.dev/s/Q5D4/ :
```
<... a ...>
...
b
```

Pattern that matches https://dev.semgrep.dev/s/gLgJ/ :
```
if <... a ...>:
  ...
b
```

I'm guessing the issue is that `<... thing ...>` can only match an expression, not a statement. The original expectation seems natural to me.

Assigning @aryx to this.

**Comment by aryx** [comment]
yes, <... ...> expect an 'expression' context, so when python parses this, it actually parses <... a ...> as an
expression statement (in other languages they have a semicolon), so matt example is really parsed internally as
```
<... a ...>;
...
b;
```
and expression statement does not match an if header.

**Comment by aryx** [comment]
We could change semgrep to have those deep expression operator go inside if/while/... headers, but I'm not sure that's what we want for other languages.

**Comment by mschwager** [comment]
> I'm guessing the issue is that <... thing ...> can only match an expression, not a statement.

Ahh, I think I didn't realize this initially. Only considering expressions makes sense to me, I just have to adjust my mental model :)

**Comment by nbrahms** [comment]
@mschwager is this issue still a problem for you, or should we close it?

**Comment by mschwager** [comment]
> @mschwager is this issue still a problem for you, or should we close it?

The issue mentioned in the description still exists, but it's not currently blocking me on anything. I'm having a tough time remembering exactly why I was trying to write this type of pattern. I think I was trying to emulate some kind of source-sanitizer-sink pattern, like: https://dev.semgrep.dev/s/PJqz.

We did some similar work for https://github.com/returntocorp/semgrep/issues/3077, and, per https://github.com/returntocorp/semgrep/issues/3077#issuecomment-839791284 (`<... $O[$ARG] ...>` now works), this issue may still be relevant. I don't have strong feelings here - I think this issue is still valid, but it's not blocking me on anything :man_shrugging:. Feel free to close if you'd like.

**Comment by emjin** [comment]
@mschwager Is this closed by taint mode?

**Comment by mschwager** [comment]
> @mschwager Is this closed by taint mode?

Hmm, possibly. Can you provide a taint rule that matches the original examples?

I guess the key confusion here is that if I have one pattern that finds statement A, and another pattern that finds statement B, and I put a `...` between those two patterns, I would expect the combined pattern to find statement A to statement B.

E.g.

>     Deep expression operator works: https://dev.semgrep.dev/s/X2Rl/
>     Subsequent function call works: https://dev.semgrep.dev/s/jxWZ/
>     Combined they don't work: https://dev.semgrep.dev/s/1LDb/

Again, not a huge deal since this isn't blocking me, but it's a minor confusion. Feel free to close if you'd like :+1:

**Comment by nbrahms** [comment]
Fixed in #3960
